crypto/dilithium/p_dilithium3_asn1.c

```diff
@@ -0,0 +1,208 @@
+#include "../internal.h"
+#include "sig_dilithium.h"
+
+
+static void dilithium3_free(EVP_PKEY *pkey) {
+  OPENSSL_free(pkey->pkey.ptr);
+  pkey->pkey.ptr = NULL;
+}
+
+static int dilithium3_set_priv_raw(EVP_PKEY *pkey, const uint8_t *in, size_t len) {
+  if (len != DILITHIUM3_PRIVATE_KEY_BYTES) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  DILITHIUM3_KEY *key = OPENSSL_malloc(sizeof(DILITHIUM3_KEY));
+  if (key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  OPENSSL_memcpy(key->priv, in, len);
+  key->has_private = 1;
+
+  dilithium3_free(pkey);
+  pkey->pkey.ptr = key;
+  return 1;
+}
+
+static int dilithium3_set_pub_raw(EVP_PKEY *pkey, const uint8_t *in, size_t len) {
+  if (len != DILITHIUM3_PUBLIC_KEY_BYTES) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_KEYBITS);
+    return 0;
+  }
+
+  DILITHIUM3_KEY *key = OPENSSL_malloc(sizeof(DILITHIUM3_KEY));
+  if (key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  OPENSSL_memcpy(key->pub, in, len);
+  key->has_private = 0;
+
+  dilithium3_free(pkey);
+  pkey->pkey.ptr = key;
+  return 1;
+}
+
+static int dilithium3_get_priv_raw(const EVP_PKEY *pkey, uint8_t *out,
+                                   size_t *out_len) {
+  const DILITHIUM3_KEY *key = pkey->pkey.ptr;
+  if (!key->has_private) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+
+  if (out == NULL) {
+    *out_len = DILITHIUM3_PRIVATE_KEY_BYTES;
+    return 1;
+  }
+
+  if (*out_len < DILITHIUM3_PRIVATE_KEY_BYTES) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  OPENSSL_memcpy(out, key->priv, DILITHIUM3_PRIVATE_KEY_BYTES);
+  *out_len = DILITHIUM3_PRIVATE_KEY_BYTES;
+  return 1;
+}
+
+static int dilithium3_get_pub_raw(const EVP_PKEY *pkey, uint8_t *out,
+                                  size_t *out_len) {
+  const DILITHIUM3_KEY *key = pkey->pkey.ptr;
+  if (out == NULL) {
+    *out_len = DILITHIUM3_PUBLIC_KEY_BYTES;
+    return 1;
+  }
+
+  if (*out_len < DILITHIUM3_PUBLIC_KEY_BYTES) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  OPENSSL_memcpy(out, key->pub, DILITHIUM3_PUBLIC_KEY_BYTES);
+  *out_len = DILITHIUM3_PUBLIC_KEY_BYTES;
+  return 1;
+}
+
+static int dilithium3_pub_decode(EVP_PKEY *out, CBS *params, CBS *key) {
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
+  // The parameters must be omitted.
+  if (CBS_len(params) != 0) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  return dilithium3_set_pub_raw(out, CBS_data(key), CBS_len(key));
+}
+
+static int dilithium3_pub_encode(CBB *out, const EVP_PKEY *pkey) {
+  const DILITHIUM3_KEY *key = pkey->pkey.ptr;
+
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
+  // TODO: finalize this definition - OCTETSTRING to BITSTRING conversion.
```

Comment:
@[dkostic](https://github.com/dkostic) dkostic [27 days ago](https://github.com/awslabs/aws-lc/pull/637/files#r1001988322)
is this TODO for aws-lc or for the draft? :)

Reply:
Kind of both, the draft will come first, but it's to clarify in this code that these are not final.